Move the introduction more towards regression instead of fitting. (#30489)
idaholab/moose: tutorials/darcy_thermo_mech/doc/content/workshop/numerical/fem_overview.md

!---

## Function Approximation

To introduce the concept of FEM, consider a polynomial regression exercise. We can
determine a polynomial function in a:

- +Discrete sense+: Using sampled locations with associated values, we want our polynomial function to be as close to the given data as possible.
- +Continuous sense+: Using a complicated function, we want to have our polynomial function as close to the given function as possible.

Let us take the polynomial in the following form:

!equation
f(x) = \sum_{i=0}^d c_i x^i

> where $c_i$ are scalar coefficients (expansion coefficients) and $x^i$ are "basis functions". Thus, the

Reviewer: ```suggestion
where $c_i$ are scalar coefficients (expansion coefficients) and monomials $x^i$ are "basis functions". Thus, the
```

Just to make it clear that, despite the new special terminology, we're still looking at an exponent here rather than an index?

             layout={'margin':{'t':10,'b':40,'r':20},
                     'xaxis':{'title':'x','range':[0,1.2],'ticks':'outside','gridcolor':'lightgrey'},
                     'yaxis':{'title':'f(x)','range':[0,1.0],'ticks':'outside','gridcolor':'lightgrey'},
                     'legend':{'x':0.05,'y':0.95,'borderwidth':1},
                     'shapes':[{'type':'rect','x0':0,'y0':0,'x1':1,'y1':1,'xref':'paper','yref':'paper'}]}

!style-end!

!---

## Polynomial Example (Continuous)

!row!
!col! width=49%

> We want to minimize the squared distance between the knwon function ($sin(x)$)

Reviewer 2: ```suggestion
We want to minimize the squared distance between the known function ($sin(x)$)
```

and the polynomial approximate ($f(x)=\sum_{0}^2 c_i x^i$) on a given domain $[0.25,1.0]$:

!equation
(sin(x)-f(x))^2.

The cumulative squared distance can be expressed as:

!equation
R = \int\limits_{0.25}^{1.0} (sin(x)-\sum_{0}^2 c_i x^i)^2 dx = ||(sin(x)-\sum_{0}^2 c_i x^i)||_2^2,

which we would like to minimize in a least-squares sense, with respect to
the coefficients in $f(x)$ (entries of $\vec{c}$):

!equation
\frac{\partial R}{\partial \vec{c}} = 0,

!col-end!

!col! width=1%

$\quad$

!col-end!

!col! width=49%

!alert! note title=Solving this problem

> We can move the derivative with recpect to $c_i$ into the integral:

Reviewer 2: ```suggestion
We can move the derivative with respect to $c_i$ into the integral:
```

!---

## Function Approximation

> To introduce the concept of FEM, consider a polynomial regression exercise. We can
> determine a polynomial function in a:

Reviewer 3: ```suggestion
To introduce the concept of FEM, consider a polynomial regression exercise. We can
search for a polynomial function that solves the following problems:
```

- +Discrete sense+: Using sampled locations with associated values, we want our polynomial function to be as close to the given data as possible.

> - +Continuous sense+: Using a complicated function, we want to have our polynomial function as close to the given function as possible.

Reviewer 3: Given instead of Using?
then given -> input

Reviewer 3: sense -> example ?

> Let us take the polynomial in the following form:

Reviewer 3: ```suggestion
Let us write the polynomial in the following form:
```

!equation
f(x) = \sum_{i=0}^d c_i x^i

where $c_i$ are scalar coefficients (expansion coefficients) and monomials
$x^i$ are "basis functions". Thus, the

> problem is to find $c_i$ such that $f(x)$ is closest the the given points or a

Reviewer 4: what if we want the function to pass through specific points?

Reviewer 3: you need at least order N (=Npoints) polynomial to do that

Reviewer 4: But then we will not use the least square method? The point I am getting at is that least square method allows for the use of lower order polynomials (less points)

Reviewer 3: > But then we will not use the least square method? 

You can still use least square if N_poly = N_points

> what if we want the function to pass through specific points?

ok I misunderstood this. Those specific points are in addition to the given points from least square

Reviewer 4: But if N_poly = N_points you can just solve Ac = y without resorting to least squares method? The solution should be identical?

Author: Resolving this, thanks for answering the question @GiudGiud !

Author: The issue is that if N_poly ! = N_points, the matrix is not invertible. So we can't just simplify. But if that condition is satisfied, then A c = b can actually be solved without doing the projection.

given function.

!---

## Polynomial Example (Discrete)

!row!
!col! width=49%

Define a set of points (Let's pick the points using $sin(x)$):

!equation

> (x_1, y_1) = (0.25,0.247) \\

Reviewer 4: Should we make a comment regarding the number of unknowns and the number of equations?

(x_2, y_2) = (0.50,0.479) \\
(x_3, y_3) = (0.75,0.682) \\
(x_4, y_4) = (1.00,0.841)

Substitute $(x_i, y_i)$ data into the polynomial model:

!equation
c_0 + c_1 x_i + c_2 x_i^2 = y_i,\quad i=1,..,4.

This leads to the following linear system for $c_i$:

!equation
\mathbf{A}
\vec{c}
=
\vec{y}

where

!style fontsize=80%
!equation
\mathbf{A} =
\begin{bmatrix}
A_{*0} | A_{*1} | A_{*2}
\end{bmatrix}
=
\begin{bmatrix}
  1 & 0.25 & 0.25^2 \\
  1 & 0.50 & 0.50^2 \\
  1 & 0.75 & 0.75^2 \\
  1 & 1.00 & 1.00^2
\end{bmatrix}, ~~
\vec{c} = \begin{bmatrix}
  c_0 \\
  c_1 \\
  c_2
\end{bmatrix}, ~~
\vec{y} =
\begin{bmatrix}
  0.247 \\
  0.479 \\
  0.682 \\
  0.841
\end{bmatrix}

!col-end!

!col! width=49%

!alert! note title=Solving this problem

We want to get the function closest to the points.
Distance in this case can be expressed by the discrete
$L^2$ norm:

!equation
R = ||\mathbf{A}\vec{c}-\vec{y}||^2_2,\quad \text{where} \quad
||\vec{v}||^2_2 = \vec{v}^T \vec{v},

We want to minimize the norm of the squared distances ("least squares"):

!equation

> \frac{\partial R}{\partial c_i} = 2 A_{*i}^T (\mathbf{A}\vec{c} - \vec{y}) = 0

Reviewer 4: How do we know that the derivative of the distance will lead us to a minimum and not a maximum?

Author: It can be proved theoretically that this will be a minimum. For this the second derivative (Hessian, A^T*A) has to be positive definite and in our case this will always be true.

Reviewer: You can look at the second derivative of the residual, AᵀA, and see that because of its form it has to be positive semidefinite (xAᵀAx = ‖Aᵀx‖² ≥ 0), so it can't be a maximum or a saddle point.

It's a little harder to see that we're at a *unique* minimum.  In this case the Hessian is positive definite, but if you construct A using fewer points than basis functions then you *do* get a merely-semidefinite Hessian and a non-unique minimum.

At some point in the future we might consider annotating these slides with notes on more details like these.  We can't fit every interesting possible question into the standard talk, but it would be nice for the speaker to have as much as background as possible on hand to respond to each audience's particular questions.

which requires the solution of this problem:

!equation

> \mathbf{A}^T \mathbf{A}\vec{c} = \mathbf{A}^T\vec{y}

Reviewer 3: are you introducing gradient descent to get here?

Reviewer 3: like the intent is to say, "we'll do one step of gradient descent, and since this is linear, we'll be there" ?

                     'yaxis':{'title':'f(x)','range':[0,1.0],'ticks':'outside','gridcolor':'lightgrey'},
                     'legend':{'x':0.05,'y':0.95,'borderwidth':1},
                     'shapes':[{'type':'rect','x0':0,'y0':0,'x1':1,'y1':1,'xref':'paper','yref':'paper'}]}

!style-end!

!---

## Polynomial Example (Continuous)

!row!
!col! width=49%

We want to minimize the squared distance between the known function ($sin(x)$)
and the polynomial approximate ($f(x)=\sum_{0}^2 c_i x^i$) on a given domain $[0.25,1.0]$:

!equation
\left(f(x) - sin(x)\right)^2.

The cumulative squared distance can be expressed as:

!equation
R = \int\limits_{0.25}^{1.0} \left(\sum_{0}^2 c_i x^i - sin(x)\right)^2 dx = ||\sum_{0}^2 c_i x^i - sin(x)||_2^2,

which we would like to minimize in a least-squares sense, with respect to
the coefficients in $f(x)$ (entries of $\vec{c}$):

!equation
\frac{\partial R}{\partial c_i} = 0, \quad i=0,1,2

!col-end!

!col! width=1%

$\quad$

!col-end!

!col! width=49%

!alert! note title=Solving this problem

We can move the derivative with respect to $c_i$ into the integral:

> 
> !equation
> \frac{\partial}{\partial c_i} = \int\limits_{0.25}^{1.0} 2 \textcolor{red}{x^i} \left(\sum_{0}^2 c_i x^i - sin(x)\right) dx,

Reviewer 4: ```suggestion

!equation
\frac{\partial R}{\partial c_i} = \int\limits_{0.25}^{1.0} 2 \textcolor{red}{x^i} \left(\sum_{0}^2 c_i x^i - sin(x)\right) dx,
```

Author: Beautiful catch!

!style! fontsize=60%

!row!

!col! width=10%

> !listing test/tests/kernels/2d_diffusion/2d_diffusion_neumannbc_test.i block=Kernels remove=Kernels/active link=False

Reviewer 3: good idea to drop the links. Probably should do this more
